src/proxy/http/metrics/handle_time.rs

```diff
@@ -0,0 +1,252 @@
+    histogram: Mutex<Histogram<latency::Us>>,
+    /// Stores the state of currently active `Tracker`s.
+    recorders: RwLock<Vec<Recorder>>,
+    /// The index of the most recently finished recorder.
+    ///
+    /// When a new recorder is needed, the recorder at this index will be used,
+    /// and that recorder's next index will be set as the head. of the free
+    /// list.
+    ///
+    /// When an active recorder completes, this will be set to its index, and
+    /// the previous value will become the freed recorder's next pointer.
+    idle_head: AtomicUsize,
+}
+
+#[derive(Debug)]
+struct Recorder {
+    /// The number of currently active `Tracker`s for this recorder.
+    ///
+    /// When a request is initially received, there will be one `Tracker` in its
+    /// `Extensions` map. If the request is cloned for retries, the `Tracker`
+    /// will be cloned, incrementing this count. Dropping a `Tracker` decrements
+    /// this count, and when it reaches 0, the handle time is recorded.
+    active_trackers: AtomicUsize,
+    /// Index of the next free recorder.
+    next_idle: AtomicUsize,
+}
+
+impl insert::Lazy<Tracker> for InsertTracker {
+    fn value(&self) -> Tracker {
+        self.0.clone().tracker()
+    }
+}
+
+// ===== impl Scope =====
+
+impl Scope {
+    pub fn new() -> Self {
+        Scope(Arc::new(Shared::new()))
+    }
+
+    pub fn layer(&self) -> insert::Layer<InsertTracker, Tracker> {
+        insert::Layer::new(InsertTracker(self.0.clone()))
+    }
+}
+
+impl FmtMetric for Scope {
+    const KIND: &'static str = <Histogram<latency::Us> as FmtMetric>::KIND;
+
+    fn fmt_metric<N: fmt::Display>(&self, f: &mut fmt::Formatter<'_>, name: N) -> fmt::Result {
+        if let Ok(hist) = self.0.histogram.lock() {
+            hist.fmt_metric(f, name)?;
+        }
+        Ok(())
+    }
+
+    fn fmt_metric_labeled<N, L>(
+        &self,
+        f: &mut fmt::Formatter<'_>,
+        name: N,
+        labels: L,
+    ) -> fmt::Result
+    where
+        N: fmt::Display,
+        L: FmtLabels,
+    {
+        if let Ok(hist) = self.0.histogram.lock() {
+            hist.fmt_metric_labeled(f, name, labels)?;
+        }
+        Ok(())
+    }
+}
+
+// ===== impl InsertTracker =====
+
+impl Clone for Tracker {
+    fn clone(&self) -> Self {
+        self.shared.clone_tracker(self.idx);
+        Self {
+            shared: self.shared.clone(),
+            idx: self.idx,
+            t0: self.t0,
+        }
+    }
+}
+
+impl Drop for Tracker {
+    fn drop(&mut self) {
+        self.shared.drop_tracker(&*self);
+    }
+}
+
+impl Shared {
+    const INITIAL_RECORDERS: usize = 32;
+
+    fn new() -> Self {
+        let mut recorders = Vec::with_capacity(Self::INITIAL_RECORDERS);
+        Self::add_recorders(&mut recorders, Self::INITIAL_RECORDERS);
+        Self {
+            histogram: Mutex::new(Histogram::default()), // TODO(eliza): should we change the bounds here?
```

Comment:
to what? I don't understand the comment

Reply:
Ah, I should probably have said "_bucket_ bounds" here, the type is just called `Bounds` in the histogram module. 

Basically, I'm not convinced that [these buckets](https://github.com/linkerd/linkerd2-proxy/blob/d315b9f6afb23eb3d0cf58577710989d1bd69944/lib/metrics/src/latency.rs#L7-L35) provide great resolution for handle times, given that it seems like a majority of them fall into the 1-2ms range; it might be worth tuning the buckets a bit.

Reply:
Yeah, I'm not convinced that those buckets are a good choice, either. Would you mind opening an issue in the proxy visibility backlog re: tuning histograms for fidelity.

Reply:
https://github.com/linkerd/linkerd2/issues/3202